File: shared/websocket_manager.py
### shared/websocket_manager.py

import asyncio
import websockets

class WebSocketManager:
    def __init__(self):
        self.clients = {}
        self.handlers = {}

    def register_handler(self, event_type, handler):
        self.handlers[event_type] = handler

    async def handle_connection(self, websocket, path):
        client_id = id(websocket)
        self.clients[client_id] = websocket
        try:
            async for message in websocket:
                await self.process_message(client_id, message)
        finally:
            del self.clients[client_id]

    async def process_message(self, client_id, message):
        data = json.loads(message)
        event_type = data.get('type')
        payload = data.get('payload')
        if handler := self.handlers.get(event_type):
            await handler(client_id, payload)

    async def broadcast(self, event_type, payload):
        message = json.dumps({'type': event_type, 'payload': payload})
        await asyncio.wait([client.send(message) for client in self.clients.values()])

    async def send_to_client(self, client_id, payload):
        if client := self.clients.get(client_id):
            await client.send(json.dumps(payload))

websocket_manager = WebSocketManager()

File: shared/websocket_manager.js
// Description: A simple WebSocket manager that allows you to connect to a WebSocket server, send messages, and subscribe to events.
// The manager keeps track of subscriptions and calls the appropriate callbacks when a message is received.
// Usage: import { websocketManager } from 'shared/websocket_manager.js';
// websocketManager.connect('ws://localhost:3000');
class WebSocketManager {
    constructor() {
        this.socket = null;
        this.subscriptions = {};
    }

    connect(url, options = {}) {
        this.socket = new WebSocket(url);

        this.socket.onopen = () => {
            console.log('WebSocket connected');
        };

        this.socket.onmessage = (event) => {
            const data = JSON.parse(event.data);
            if (this.subscriptions[data.type]) {
                this.subscriptions[data.type].forEach(cb => cb(data.payload));
            }
        };

        this.socket.onerror = (error) => {
            console.error('WebSocket error:', error);
        };

        this.socket.onclose = () => {
            console.log('WebSocket disconnected');
        };
    }

    send(message) {
        if (this.socket.readyState === WebSocket.OPEN) {
            this.socket.send(JSON.stringify(message));
            return true;
        }
        return false;
    }

    subscribe(eventType, callback) {
        const subscriptionId = `sub_${Date.now()}`;
        this.subscriptions[event_type] = this.subscriptions[event_type] || [];
        this.subscriptions[event_type].push({ id: subscriptionId, callback });
        return subscriptionId;
    }

    close() {
        if (this.socket) {
            this.socket.close();
        }
    }
}

export const websocketManager = new WebSocketManager();